File: leet_lint_code/leet_code/binary_tree_level_order_traversal_ii.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 107.

#ifndef LEETLINTCODE_LINTCODE_BINARYTREELEVELORDERTRAVERSALII_CPP
#define LEETLINTCODE_LINTCODE_BINARYTREELEVELORDERTRAVERSALII_CPP

#include <queue>
#include <vector>
#include <math.h>

class TreeNode {
 public:
   TreeNode(int val) {
     this->val = val;
     this->left = this->right = NULL;
   }

   int val;
   TreeNode *left, *right;
};

class BinaryTreeLevelOrderTraversalII {
 public:
   std::vector<std::vector<int>> levelOrderBottom(TreeNode* root) {
     std::vector<std::vector<int>> v;
     int level = 0;
     levelOrderBottomHelp(root, v, 0, level);
     return v;
   }

   void levelOrderBottomHelp(TreeNode* root, std::vector<std::vector<int>> &v, int current_level, int &level) {
     if (NULL == root) return;
     levelOrderBottomHelp(root->left, v, current_level + 1, level);
     levelOrderBottomHelp(root->right, v, current_level + 1, level);
     if (level < current_level) level = current_level;
     int index = level - current_level;
     int len = (int)v.size();
     if (len - 1 < index) {
       std::vector<int> tmp;
       v.push_back(tmp);
     }

     v[index].push_back(root->val);
     return ;
   }
};

#endif // LEETLINTCODE_LINTCODE_BINARYTREELEVELORDERTRAVERSALII_CPP

File: leet_lint_code/leet_code/java/BinaryTreeLevelOrderTraversalII.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 103.
// Date: 2019.8.14

import java.util.*;

public class BinaryTreeLevelOrderTraversalII {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        Stack<TreeNode> stack1 = new Stack<>();
        Stack<TreeNode> stack2 = new Stack<>();
        if (root == null) {
            return ans;
        }

        // 0. left to right 1. right to left
        int order = 0;
        stack2.push(root);
        while (!stack2.isEmpty()) {
            ans.add(levelOrder(stack1, stack2, order));
            stack2.addAll(stack1);
            stack1.clear();
            order++;
            order %= 2;
        }

        return ans;
    }

    List<Integer> levelOrder(Stack<TreeNode> stack1, Stack<TreeNode> stack2, int order) {
        List<Integer> levelAns = new ArrayList<>();
        if (order == 0) {
            while (!stack2.isEmpty()) {
                TreeNode top = stack2.pop();
                if (top.left != null) {
                    stack1.push(top.left);
                }

                if (top.right != null) {
                    stack1.push(top.right);
                }

                levelAns.add(top.val);
            }
        } else {
            while (!stack2.isEmpty()) {
                TreeNode top = stack2.pop();
                if (top.right != null) {
                    stack1.push(top.right);
                }

                if (top.left != null) {
                    stack1.push(top.left);
                }

                levelAns.add(top.val);
            }
        }

        stack2.clear();
        return levelAns;
    }
}
